### PR TITLE
Tests/uni v3 vault

### PR DESCRIPTION
**Pull request by AlexSaplin**

### CODE REVIEW

File: test/behaviors/integrationVault.ts [hunk 1]
```diff
@@ -0,0 +1,799 @@
...
+) {
+    const APPROVE_SELECTOR = "0x095ea7b3";
+    describe("#push", () => {
+        it("emits Push event", async () => {
+            await expect(
+                this.subject.push([this.usdc.address], [BigNumber.from(1)], [])
+            ).to.emit(this.subject, "Push");
+        });
+        it("passes when tokens transferred", async () => {
+            const result = await mintUniV3Position_USDC_WETH({
```
Comment (alexeuler) [inline test/behaviors/integrationVault.ts:42]: All uniV3 related stuff should go into UniV3.test.ts

File: test/behaviors/integrationVault.ts [hunk 2]
```diff
@@ -0,0 +1,799 @@
...
+                [],
+            ];
+            const amounts = await this.subject.callStatic.push(...args);
+            expect(amounts[0]).to.deep.equal(
+                BigNumber.from(10).pow(6).mul(3000)
+            );
+        });
+
+        describe("edge cases", () => {
+            it("reverts when vault's nft is 0", async () => {
```
[the remark this answers is not in the document]
Comment (alexeuler) [reply test/behaviors/integrationVault.ts:67]: Also let's specifty the exception in the spec
`reverts with '${Exceptions.INIT}'`

File: test/behaviors/integrationVault.ts [hunk 3]
```diff
@@ -0,0 +1,799 @@
...
+    {}: {}
+) {
+    const APPROVE_SELECTOR = "0x095ea7b3";
+    describe("#push", () => {
+        it("emits Push event", async () => {
+            await expect(
+                this.subject.push([this.usdc.address], [BigNumber.from(1)], [])
+            ).to.emit(this.subject, "Push");
+        });
+        it("passes when tokens transferred", async () => {
```
Comment (alexeuler) [inline test/behaviors/integrationVault.ts:41]: It's better to name happy case to what it does exactly. E.g. `pushes tokens to the underlying protocol`

File: test/behaviors/integrationVault.ts [hunk 4]
```diff
@@ -0,0 +1,799 @@
...
+            it("reverts when tokens are not unique", async () => {
+                await expect(
+                    this.subject.push(
+                        [this.usdc.address, this.usdc.address],
+                        [BigNumber.from(1), BigNumber.from(1)],
+                        []
+                    )
+                ).to.be.revertedWith(Exceptions.INVARIANT);
+            });
+            it("reverts when tokens not sorted nor unique", async () => {
```
Comment (alexeuler) [inline test/behaviors/integrationVault.ts:130]: when tokens are not unique

File: test/behaviors/integrationVault.ts [hunk 5]
```diff
@@ -0,0 +1,799 @@
...
+                            .pull(
+                                this.erc20Vault.address,
+                                [this.usdc.address],
+                                [BigNumber.from(1)],
+                                []
+                            )
+                    ).to.not.be.reverted;
+                });
+            });
+            it("not allowed: admin", async () => {
```
Comment (alexeuler) [inline test/behaviors/integrationVault.ts:549]: Just as a common style let's use forbidden: admin here

File: test/behaviors/integrationVault.ts [hunk 6]
```diff
@@ -0,0 +1,799 @@
...
+                                []
+                            )
+                    ).to.be.revertedWith(Exceptions.FORBIDDEN);
+                });
+            });
+        });
+    });
+
+    describe("#isValidSignature", () => {
+        describe.only("edge cases:", () => {
```
Comment (alexeuler) [inline test/behaviors/integrationVault.ts:579]: watch out before merge :)

**Review by alexeuler — CHANGES_REQUESTED** [review]

File: test/behaviors/integrationVault.ts (hunk 4 above)
Comment (AlexSaplin) [reply test/behaviors/integrationVault.ts:130]: "when tokens are not unique" test goes above of that

**Review by alexeuler — APPROVED** [review]
